src/sync/app_credentials.hpp

```diff
@@ -0,0 +1,112 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_APP_CREDENTIALS_HPP
+#define REALM_APP_CREDENTIALS_HPP
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace realm {
+namespace app {
+
+typedef std::string IdentityProvider;
+typedef std::string AppCredentialsToken;
+
+/// The username/password identity provider. User accounts are handled by the Realm Object Server directly without the
+/// involvement of a third-party identity provider.
+extern IdentityProvider const IdentityProviderUsernamePassword;
+
+/// A Facebook account as an identity provider.
+extern IdentityProvider const IdentityProviderFacebook;
+
+/// A Google account as an identity provider.
+extern IdentityProvider const IdentityProviderGoogle;
+
+/// A JSON Web Token as an identity provider.
+extern IdentityProvider const IdentityProviderJWT;
+
+/// An Anonymous account as an identity provider.
+extern IdentityProvider const IdentityProviderAnonymous;
+
+/// A Google account as an identity provider.
+extern IdentityProvider const IdentityProviderApple;
+
+enum class AuthProvider {
+    ANONYMOUS,
+    FACEBOOK,
+    APPLE,
+    USERNAME_PASSWORD
+};
+
+IdentityProvider provider_type_from_enum(AuthProvider provider);
+
+/**
+ Opaque credentials representing a specific Realm Object Server user.
+ */
+struct AppCredentials {
+public:
+    /**
+     Construct and return credentials from a Facebook account token.
+     */
+    static std::shared_ptr<AppCredentials> facebook(const AppCredentialsToken access_token);
```

Comment:
There doesn't appear to be any reason for these to be returning a `std::shared_ptr<AppCredentials>` rather than just returning a `AppCredentials` (and similarly `login_with_credentials` doesn't have any reason to be taking a `std::shared_ptr<AppCredentials>` rather than a const ref to a `AppCredentials`).